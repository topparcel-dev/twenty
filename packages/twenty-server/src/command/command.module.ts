import { Module } from '@nestjs/common';

import { AppModule } from 'src/app.module';
import { DatabaseCommandModule } from 'src/database/commands/database-command.module';
import { WorkspaceCleanerModule } from 'src/engine/workspace-manager/workspace-cleaner/workspace-cleaner.module';
import { WorkspaceHealthCommandModule } from 'src/engine/workspace-manager/workspace-health/commands/workspace-health-command.module';
import { WorkspaceMigrationRunnerCommandsModule } from 'src/engine/workspace-manager/workspace-migration-runner/commands/workspace-migration-runner-commands.module';
import { WorkspaceSyncMetadataCommandsModule } from 'src/engine/workspace-manager/workspace-sync-metadata/commands/workspace-sync-metadata-commands.module';

@Module({
  imports: [
    AppModule,
    WorkspaceSyncMetadataCommandsModule,
    DatabaseCommandModule,
    WorkspaceCleanerModule,
    WorkspaceHealthCommandModule,
    WorkspaceMigrationRunnerCommandsModule,
  ],
})
export class CommandModule {}
