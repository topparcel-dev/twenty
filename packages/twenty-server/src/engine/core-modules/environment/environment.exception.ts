import { CustomException } from 'src/utils/custom-exception';

export class EnvironmentException extends CustomException {
  code: EnvironmentExceptionCode;
  constructor(message: string, code: EnvironmentExceptionCode) {
    super(message, code);
  }
}

export enum EnvironmentExceptionCode {
  ENVIRONMENT_VARIABLES_NOT_FOUND = 'ENVIRONMENT_VARIABLES_NOT_FOUND',
}
