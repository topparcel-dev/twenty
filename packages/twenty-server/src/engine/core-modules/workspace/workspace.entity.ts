import { Field, ObjectType, registerEnumType } from '@nestjs/graphql';

import { IDField, UnPagedRelation } from '@ptc-org/nestjs-query-graphql';
import {
  Column,
  CreateDateColumn,
  Entity,
  OneToMany,
  PrimaryGeneratedColumn,
  Relation,
  UpdateDateColumn,
} from 'typeorm';

import { UUIDScalarType } from 'src/engine/api/graphql/workspace-schema-builder/graphql-types/scalars';
import { AppToken } from 'src/engine/core-modules/app-token/app-token.entity';
import { BillingCustomer } from 'src/engine/core-modules/billing/entities/billing-customer.entity';
import { BillingEntitlement } from 'src/engine/core-modules/billing/entities/billing-entitlement.entity';
import { BillingSubscription } from 'src/engine/core-modules/billing/entities/billing-subscription.entity';
import { FeatureFlagEntity } from 'src/engine/core-modules/feature-flag/feature-flag.entity';
import { KeyValuePair } from 'src/engine/core-modules/key-value-pair/key-value-pair.entity';
import { PostgresCredentials } from 'src/engine/core-modules/postgres-credentials/postgres-credentials.entity';
import { WorkspaceSSOIdentityProvider } from 'src/engine/core-modules/sso/workspace-sso-identity-provider.entity';
import { UserWorkspace } from 'src/engine/core-modules/user-workspace/user-workspace.entity';
import { User } from 'src/engine/core-modules/user/user.entity';

export enum WorkspaceActivationStatus {
  ONGOING_CREATION = 'ONGOING_CREATION',
  PENDING_CREATION = 'PENDING_CREATION',
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
}

registerEnumType(WorkspaceActivationStatus, {
  name: 'WorkspaceActivationStatus',
});

@Entity({ name: 'workspace', schema: 'core' })
@ObjectType('Workspace')
@UnPagedRelation('featureFlags', () => FeatureFlagEntity, { nullable: true })
@UnPagedRelation('billingSubscriptions', () => BillingSubscription, {
  nullable: true,
})
@UnPagedRelation('billingEntitlements', () => BillingEntitlement, {
  nullable: true,
})
@UnPagedRelation('billingCustomers', () => BillingCustomer, {
  nullable: true,
})
export class Workspace {
  @IDField(() => UUIDScalarType)
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  domainName?: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  displayName?: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  logo?: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  inviteHash?: string;

  @Field({ nullable: true })
  @Column({ nullable: true, type: 'timestamptz' })
  deletedAt?: Date;

  @Field()
  @CreateDateColumn({ type: 'timestamptz' })
  createdAt: Date;

  @Field()
  @UpdateDateColumn({ type: 'timestamptz' })
  updatedAt: Date;

  @OneToMany(() => AppToken, (appToken) => appToken.workspace, {
    cascade: true,
  })
  appTokens: Relation<AppToken[]>;

  @OneToMany(() => KeyValuePair, (keyValuePair) => keyValuePair.workspace, {
    cascade: true,
  })
  keyValuePairs: Relation<KeyValuePair[]>;

  @OneToMany(() => User, (user) => user.defaultWorkspace)
  users: Relation<User[]>;

  @OneToMany(() => UserWorkspace, (userWorkspace) => userWorkspace.workspace, {
    onDelete: 'CASCADE',
  })
  workspaceUsers: Relation<UserWorkspace[]>;

  @Field()
  @Column({ default: true })
  allowImpersonation: boolean;

  @Field()
  @Column({ default: true })
  isPublicInviteLinkEnabled: boolean;

  @OneToMany(() => FeatureFlagEntity, (featureFlag) => featureFlag.workspace)
  featureFlags: Relation<FeatureFlagEntity[]>;

  @Field({ nullable: true })
  workspaceMembersCount: number;

  @Field(() => WorkspaceActivationStatus)
  @Column({
    type: 'enum',
    enum: WorkspaceActivationStatus,
    default: WorkspaceActivationStatus.INACTIVE,
  })
  activationStatus: WorkspaceActivationStatus;

  @OneToMany(
    () => PostgresCredentials,
    (postgresCredentials) => postgresCredentials.workspace,
  )
  allPostgresCredentials: Relation<PostgresCredentials[]>;

  @OneToMany(
    () => WorkspaceSSOIdentityProvider,
    (workspaceSSOIdentityProviders) => workspaceSSOIdentityProviders.workspace,
  )
  workspaceSSOIdentityProviders: Relation<WorkspaceSSOIdentityProvider[]>;

  @Field()
  @Column({ default: 1 })
  metadataVersion: number;

  @Field()
  @Column({ default: '' })
  databaseUrl: string;

  @Field()
  @Column({ default: '' })
  databaseSchema: string;

  @Field()
  @Column()
  subdomain: string;

  @Field()
  @Column({ default: true })
  isGoogleAuthEnabled: boolean;

  @Field()
  @Column({ default: true })
  isPasswordAuthEnabled: boolean;

  @Field()
  @Column({ default: false })
  isMicrosoftAuthEnabled: boolean;
}
