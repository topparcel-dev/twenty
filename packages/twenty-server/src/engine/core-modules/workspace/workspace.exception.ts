import { CustomException } from 'src/utils/custom-exception';

export class WorkspaceException extends CustomException {
  constructor(message: string, code: WorkspaceExceptionCode) {
    super(message, code);
  }
}

export enum WorkspaceExceptionCode {
  SUBDOMAIN_NOT_FOUND = 'SUBDOMAIN_NOT_FOUND',
  SUBDOMAIN_ALREADY_TAKEN = 'SUBDOMAIN_ALREADY_TAKEN',
  WORKSPACE_NOT_FOUND = 'WORKSPACE_NOT_FOUND',
}
