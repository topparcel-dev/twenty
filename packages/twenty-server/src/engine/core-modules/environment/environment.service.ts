/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { ENVIRONMENT_VARIABLES_MASKING_CONFIG } from 'src/engine/core-modules/environment/constants/environment-variables-masking-config';
import { EnvironmentVariablesMetadataOptions } from 'src/engine/core-modules/environment/decorators/environment-variables-metadata.decorator';
import { EnvironmentVariablesMaskingStrategies } from 'src/engine/core-modules/environment/enums/environment-variables-masking-strategies.enum';
import { EnvironmentVariables } from 'src/engine/core-modules/environment/environment-variables';
import { environmentVariableMaskSensitiveData } from 'src/engine/core-modules/environment/utils/environment-variable-mask-sensitive-data.util';
import { TypedReflect } from 'src/utils/typed-reflect';

@Injectable()
export class EnvironmentService {
  constructor(private readonly configService: ConfigService) {}

  get<T extends keyof EnvironmentVariables>(key: T): EnvironmentVariables[T] {
    return this.configService.get<EnvironmentVariables[T]>(
      key,
      new EnvironmentVariables()[key],
    );
  }

  getAll(): Record<
    string,
    {
      value: EnvironmentVariables[keyof EnvironmentVariables];
      metadata: EnvironmentVariablesMetadataOptions;
    }
  > {
    const result: Record<
      string,
      {
        value: EnvironmentVariables[keyof EnvironmentVariables];
        metadata: EnvironmentVariablesMetadataOptions;
      }
    > = {};

    const envVars = new EnvironmentVariables();
    const metadata =
      TypedReflect.getMetadata('environment-variables', EnvironmentVariables) ??
      {};

    Object.entries(metadata).forEach(([key, envMetadata]) => {
      let value =
        this.configService.get(key) ??
        envVars[key as keyof EnvironmentVariables] ??
        '';

      if (
        typeof value === 'string' &&
        key in ENVIRONMENT_VARIABLES_MASKING_CONFIG
      ) {
        const varMaskingConfig =
          ENVIRONMENT_VARIABLES_MASKING_CONFIG[
            key as keyof typeof ENVIRONMENT_VARIABLES_MASKING_CONFIG
          ];
        const options =
          varMaskingConfig.strategy ===
          EnvironmentVariablesMaskingStrategies.LAST_N_CHARS
            ? { chars: varMaskingConfig.chars }
            : undefined;

        value = environmentVariableMaskSensitiveData(
          value,
          varMaskingConfig.strategy,
          { ...options, variableName: key },
        );
      }

      result[key] = {
        value,
        metadata: envMetadata,
      };
    });

    return result;
  }
}
