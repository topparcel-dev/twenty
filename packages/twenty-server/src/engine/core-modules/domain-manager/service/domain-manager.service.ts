import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';

import { Repository } from 'typeorm';

import { EnvironmentService } from 'src/engine/core-modules/environment/environment.service';
import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
import {
  WorkspaceException,
  WorkspaceExceptionCode,
} from 'src/engine/core-modules/workspace/workspace.exception';
import { getDomainNameByEmail } from 'src/utils/get-domain-name-by-email';
import { isDefined } from 'src/utils/is-defined';
import { isWorkEmail } from 'src/utils/is-work-email';

@Injectable()
export class DomainManagerService {
  constructor(
    @InjectRepository(Workspace, 'core')
    private readonly workspaceRepository: Repository<Workspace>,
    private readonly environmentService: EnvironmentService,
  ) {}

  getFrontUrl() {
    let baseUrl: URL;
    const frontPort = this.environmentService.get('FRONT_PORT');
    const frontDomain = this.environmentService.get('FRONT_DOMAIN');
    const frontProtocol = this.environmentService.get('FRONT_PROTOCOL');

    const serverUrl = this.environmentService.get('SERVER_URL');

    if (!frontDomain) {
      baseUrl = new URL(serverUrl);
    } else {
      baseUrl = new URL(`${frontProtocol}://${frontDomain}`);
    }

    if (frontPort) {
      baseUrl.port = frontPort.toString();
    }

    if (frontProtocol) {
      baseUrl.protocol = frontProtocol;
    }

    return baseUrl;
  }

  getBaseUrl(): URL {
    const baseUrl = this.getFrontUrl();

    if (
      this.environmentService.get('IS_MULTIWORKSPACE_ENABLED') &&
      this.environmentService.get('DEFAULT_SUBDOMAIN')
    ) {
      baseUrl.hostname = `${this.environmentService.get('DEFAULT_SUBDOMAIN')}.${baseUrl.hostname}`;
    }

    return baseUrl;
  }

  buildWorkspaceURL({
    subdomain,
    pathname,
    searchParams,
  }: {
    subdomain?: string;
    pathname?: string;
    searchParams?: Record<string, string | number>;
  }) {
    const url = this.getBaseUrl();

    if (
      this.environmentService.get('IS_MULTIWORKSPACE_ENABLED') &&
      !subdomain
    ) {
      throw new Error('subdomain is required when multiworkspace is enable');
    }

    if (
      subdomain &&
      subdomain.length > 0 &&
      this.environmentService.get('IS_MULTIWORKSPACE_ENABLED')
    ) {
      url.hostname = url.hostname.replace(
        this.environmentService.get('DEFAULT_SUBDOMAIN'),
        subdomain,
      );
    }

    if (pathname) {
      url.pathname = pathname;
    }

    if (searchParams) {
      Object.entries(searchParams).forEach(([key, value]) => {
        if (isDefined(value)) {
          url.searchParams.set(key, value.toString());
        }
      });
    }

    return url;
  }

  getWorkspaceSubdomainByOrigin = (origin: string) => {
    const { hostname: originHostname } = new URL(origin);

    const frontDomain = this.getFrontUrl().hostname;

    const subdomain = originHostname.replace(`.${frontDomain}`, '');

    if (this.isDefaultSubdomain(subdomain)) {
      return;
    }

    return subdomain;
  };

  isDefaultSubdomain(subdomain: string) {
    return subdomain === this.environmentService.get('DEFAULT_SUBDOMAIN');
  }

  computeRedirectErrorUrl({
    errorMessage,
    subdomain,
  }: {
    errorMessage: string;
    subdomain?: string;
  }) {
    const url = this.buildWorkspaceURL({
      subdomain,
      pathname: '/verify',
      searchParams: { errorMessage },
    });

    return url.toString();
  }

  async getDefaultWorkspace() {
    if (!this.environmentService.get('IS_MULTIWORKSPACE_ENABLED')) {
      const workspaces = await this.workspaceRepository.find({
        order: {
          createdAt: 'DESC',
        },
        relations: ['workspaceSSOIdentityProviders'],
      });

      if (workspaces.length > 1) {
        // TODO AMOREAUX: this logger is trigger twice and the second time the message is undefined for an unknown reason
        Logger.warn(
          `In single-workspace mode, there should be only one workspace. Today there are ${workspaces.length} workspaces`,
        );
      }

      return workspaces[0];
    }

    throw new Error(
      'Default workspace not exist when multi-workspace is enabled',
    );
  }

  async getWorkspaceByOrigin(origin: string) {
    try {
      if (!this.environmentService.get('IS_MULTIWORKSPACE_ENABLED')) {
        return this.getDefaultWorkspace();
      }

      const subdomain = this.getWorkspaceSubdomainByOrigin(origin);

      if (!isDefined(subdomain)) return;

      const workspace = await this.workspaceRepository.findOne({
        where: { subdomain },
        relations: ['workspaceSSOIdentityProviders'],
      });

      return workspace;
    } catch (e) {
      throw new WorkspaceException(
        'Workspace not found',
        WorkspaceExceptionCode.SUBDOMAIN_NOT_FOUND,
      );
    }
  }

  private generateRandomSubdomain(): string {
    const prefixes = [
      'cool',
      'smart',
      'fast',
      'bright',
      'shiny',
      'happy',
      'funny',
      'clever',
      'brave',
      'kind',
      'gentle',
      'quick',
      'sharp',
      'calm',
      'silent',
      'lucky',
      'fierce',
      'swift',
      'mighty',
      'noble',
      'bold',
      'wise',
      'eager',
      'joyful',
      'glad',
      'zany',
      'witty',
      'bouncy',
      'graceful',
      'colorful',
    ];
    const suffixes = [
      'raccoon',
      'panda',
      'whale',
      'tiger',
      'dolphin',
      'eagle',
      'penguin',
      'owl',
      'fox',
      'wolf',
      'lion',
      'bear',
      'hawk',
      'shark',
      'sparrow',
      'moose',
      'lynx',
      'falcon',
      'rabbit',
      'hedgehog',
      'monkey',
      'horse',
      'koala',
      'kangaroo',
      'elephant',
      'giraffe',
      'panther',
      'crocodile',
      'seal',
      'octopus',
    ];

    const randomPrefix = prefixes[Math.floor(Math.random() * prefixes.length)];
    const randomSuffix = suffixes[Math.floor(Math.random() * suffixes.length)];

    return `${randomPrefix}-${randomSuffix}`;
  }

  private getSubdomainNameByEmail(email?: string) {
    if (!isDefined(email) || !isWorkEmail(email)) return;

    return getDomainNameByEmail(email);
  }

  private getSubdomainNameByDisplayName(displayName?: string) {
    if (!isDefined(displayName)) return;
    const displayNameWords = displayName.match(/(\w| |\d)+/g);

    if (displayNameWords) {
      return displayNameWords.join('-').replace(/ /g, '').toLowerCase();
    }
  }

  async generateSubdomain(params?: { email?: string; displayName?: string }) {
    const subdomain =
      this.getSubdomainNameByEmail(params?.email) ??
      this.getSubdomainNameByDisplayName(params?.displayName) ??
      this.generateRandomSubdomain();

    const existingWorkspaceCount = await this.workspaceRepository.countBy({
      subdomain,
    });

    return `${subdomain}${existingWorkspaceCount > 0 ? `-${Math.random().toString(36).substring(2, 10)}` : ''}`;
  }
}
