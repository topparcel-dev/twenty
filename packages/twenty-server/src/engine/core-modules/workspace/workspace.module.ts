import { Module } from '@nestjs/common';

import { NestjsQueryGraphQLModule } from '@ptc-org/nestjs-query-graphql';
import { NestjsQueryTypeOrmModule } from '@ptc-org/nestjs-query-typeorm';

import { TypeORMModule } from 'src/database/typeorm/typeorm.module';
import { BillingModule } from 'src/engine/core-modules/billing/billing.module';
import { FeatureFlagModule } from 'src/engine/core-modules/feature-flag/feature-flag.module';
import { FileUploadModule } from 'src/engine/core-modules/file/file-upload/file-upload.module';
import { FileModule } from 'src/engine/core-modules/file/file.module';
import { OnboardingModule } from 'src/engine/core-modules/onboarding/onboarding.module';
import { UserWorkspace } from 'src/engine/core-modules/user-workspace/user-workspace.entity';
import { UserWorkspaceModule } from 'src/engine/core-modules/user-workspace/user-workspace.module';
import { User } from 'src/engine/core-modules/user/user.entity';
import { WorkspaceWorkspaceMemberListener } from 'src/engine/core-modules/workspace/workspace-workspace-member.listener';
import { WorkspaceResolver } from 'src/engine/core-modules/workspace/workspace.resolver';
import { DataSourceModule } from 'src/engine/metadata-modules/data-source/data-source.module';
import { WorkspaceMetadataCacheModule } from 'src/engine/metadata-modules/workspace-metadata-cache/workspace-metadata-cache.module';
import { WorkspaceManagerModule } from 'src/engine/workspace-manager/workspace-manager.module';
import { DomainManagerModule } from 'src/engine/core-modules/domain-manager/domain-manager.module';
import { TokenModule } from 'src/engine/core-modules/auth/token/token.module';

import { workspaceAutoResolverOpts } from './workspace.auto-resolver-opts';
import { Workspace } from './workspace.entity';

import { WorkspaceService } from './services/workspace.service';

@Module({
  imports: [
    TypeORMModule,
    NestjsQueryGraphQLModule.forFeature({
      imports: [
        DomainManagerModule,
        BillingModule,
        FileModule,
        TokenModule,
        FileUploadModule,
        WorkspaceMetadataCacheModule,
        NestjsQueryTypeOrmModule.forFeature(
          [User, Workspace, UserWorkspace],
          'core',
        ),
        UserWorkspaceModule,
        WorkspaceManagerModule,
        FeatureFlagModule,
        DataSourceModule,
        OnboardingModule,
        TypeORMModule,
      ],
      services: [WorkspaceService],
      resolvers: workspaceAutoResolverOpts,
    }),
  ],
  exports: [WorkspaceService],
  providers: [
    WorkspaceResolver,
    WorkspaceService,
    WorkspaceWorkspaceMemberListener,
  ],
})
export class WorkspaceModule {}
