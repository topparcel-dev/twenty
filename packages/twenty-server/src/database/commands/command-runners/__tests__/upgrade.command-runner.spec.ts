import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { SemVer } from 'semver';
import { Repository } from 'typeorm';

import { EnvironmentVariables } from 'src/engine/core-modules/environment/environment-variables';
import { EnvironmentService } from 'src/engine/core-modules/environment/environment.service';
import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
import { WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';
import { TwentyORMGlobalManager } from 'src/engine/twenty-orm/twenty-orm-global.manager';
import { SyncWorkspaceMetadataCommand } from 'src/engine/workspace-manager/workspace-sync-metadata/commands/sync-workspace-metadata.command';
import { UpgradeCommandRunner } from '../upgrade.command-runner';

class TestUpgradeCommandRunnerV1 extends UpgradeCommandRunner {
  fromVersion = new SemVer('1.0.0');

  public override async beforeSyncMetadataUpgradeCommandsToRun(): Promise<void> {
    return;
  }

  public override async afterSyncMetadataUpgradeCommandsToRun(): Promise<void> {
    return;
  }
}

class InvalidVersionUpgradeCommandRunner extends UpgradeCommandRunner {
  fromVersion = new SemVer('invalid');

  protected async beforeSyncMetadataUpgradeCommandsToRun(): Promise<void> {
    return;
  }

  protected async afterSyncMetadataUpgradeCommandsToRun(): Promise<void> {
    return;
  }
}

class TestUpgradeCommandRunnerV2 extends UpgradeCommandRunner {
  fromVersion = new SemVer('2.0.0');

  protected async beforeSyncMetadataUpgradeCommandsToRun(): Promise<void> {
    return;
  }

  protected async afterSyncMetadataUpgradeCommandsToRun(): Promise<void> {
    return;
  }
}

type CommandRunnerValues =
  | typeof TestUpgradeCommandRunnerV1
  | typeof TestUpgradeCommandRunnerV2
  | typeof InvalidVersionUpgradeCommandRunner;

describe('fromVersion validation', () => {
  it('should fail if fromVersion is not a valid semver', () => {
    expect(() => {
      new InvalidVersionUpgradeCommandRunner(
        {} as WorkspaceRepository<Workspace>,
        {} as EnvironmentService,
        {} as TwentyORMGlobalManager,
        {} as SyncWorkspaceMetadataCommand,
      );
    }).toThrowErrorMatchingInlineSnapshot(`"Invalid Version: invalid"`);
  });
});

const genereateMockWorkspace = (overrides?: Partial<Workspace>) =>
  ({
    id: 'workspace-id',
    version: '1.0.0',
    createdAt: new Date(),
    updatedAt: new Date(),
    allowImpersonation: false,
    isPublicInviteLinkEnabled: false,
    displayName: 'Test Workspace',
    domainName: 'test',
    inviteHash: 'hash',
    logo: null,
    deletedAt: null,
    activationStatus: 'active',
    workspaceMembersCount: 1,
    ...overrides,
  }) as Workspace;

type BuildUpgradeCommandModuleArgs = {
  workspaces: Workspace[];
  appVersion: string | null;
  commandRunner: CommandRunnerValues;
};
const buildUpgradeCommandModule = async ({
  workspaces,
  appVersion,
  commandRunner,
}: BuildUpgradeCommandModuleArgs) => {
  const module: TestingModule = await Test.createTestingModule({
    providers: [
      commandRunner,
      {
        provide: getRepositoryToken(Workspace, 'core'),
        useValue: {
          findOneByOrFail: jest
            .fn()
            .mockImplementation((args) =>
              workspaces.find((el) => el.id === args.id),
            ),
          update: jest.fn(),
          find: jest.fn().mockResolvedValue(workspaces),
        },
      },
      {
        provide: EnvironmentService,
        useValue: {
          get: jest
            .fn()
            .mockImplementation((key: keyof EnvironmentVariables) => {
              switch (key) {
                case 'APP_VERSION': {
                  return appVersion;
                }
                default: {
                  return;
                }
              }
            }),
        },
      },
      {
        provide: TwentyORMGlobalManager,
        useValue: {
          connect: jest.fn(),
          destroyDataSourceForWorkspace: jest.fn(),
          getDataSourceForWorkspace: jest.fn(),
        },
      },
      {
        provide: SyncWorkspaceMetadataCommand,
        useValue: {
          runOnWorkspace: jest.fn(),
        },
      },
    ],
  }).compile();

  return module;
};

describe('UpgradeCommandRunner', () => {
  let upgradeCommandRunner: TestUpgradeCommandRunnerV1;
  let workspaceRepository: Repository<Workspace>;
  let syncWorkspaceMetadataCommand: jest.Mocked<SyncWorkspaceMetadataCommand>;
  let afterSyncMetadataUpgradeCommandsToRun: jest.SpyInstance; //TODO improve typing ?
  let beforeSyncMetadataUpgradeCommandsToRun: jest.SpyInstance; //TODO improve typing ?
  let errors: Error[] = [];

  type BuildModuleAndSetupSpiesArgs = {
    numberOfWorkspace?: number;
    workspaceOverride?: Partial<Workspace>;
    appVersion?: string | null;
    commandRunner?: CommandRunnerValues;
  };
  const buildModuleAndSetupSpies = async ({
    numberOfWorkspace = 1,
    workspaceOverride,
    commandRunner = TestUpgradeCommandRunnerV1,
    appVersion = '2.0.0',
  }: BuildModuleAndSetupSpiesArgs) => {
    const workspaces = Array.from({ length: numberOfWorkspace }, (_v, index) =>
      genereateMockWorkspace({
        id: `workspace_${index}`,
        ...workspaceOverride,
      }),
    );
    const module = await buildUpgradeCommandModule({
      commandRunner,
      appVersion,
      workspaces,
    });

    errors = [];
    upgradeCommandRunner = module.get(commandRunner);
    beforeSyncMetadataUpgradeCommandsToRun = jest.spyOn(
      upgradeCommandRunner,
      'beforeSyncMetadataUpgradeCommandsToRun',
    );
    afterSyncMetadataUpgradeCommandsToRun = jest.spyOn(
      upgradeCommandRunner,
      'afterSyncMetadataUpgradeCommandsToRun',
    );
    jest.spyOn(upgradeCommandRunner, 'runOnWorkspace');

    workspaceRepository = module.get<Repository<Workspace>>(
      getRepositoryToken(Workspace, 'core'),
    );
    syncWorkspaceMetadataCommand = module.get(SyncWorkspaceMetadataCommand);
  };

  it('should run upgrade over several workspaces', async () => {
    const numberOfWorkspace = 42;
    await buildModuleAndSetupSpies({
      numberOfWorkspace,
    });
    const passedParams = [];
    const options = {};
    await upgradeCommandRunner.run(passedParams, options);

    [
      upgradeCommandRunner.runOnWorkspace,
      upgradeCommandRunner.beforeSyncMetadataUpgradeCommandsToRun,
      upgradeCommandRunner.afterSyncMetadataUpgradeCommandsToRun,
      syncWorkspaceMetadataCommand.runOnWorkspace,
      workspaceRepository.update,
    ].forEach((fn) => expect(fn).toHaveBeenCalledTimes(numberOfWorkspace));
    expect(upgradeCommandRunner.migrationReport).toMatchInlineSnapshot();
  });

  it('should run syncMetadataCommand betwen beforeSyncMetadataUpgradeCommandsToRun and afterSyncMetadataUpgradeCommandsToRun', async () => {
    await buildModuleAndSetupSpies({});
    const passedParams = [];
    const options = {};
    await upgradeCommandRunner.run(passedParams, options);

    [
      upgradeCommandRunner.runOnWorkspace,
      upgradeCommandRunner.beforeSyncMetadataUpgradeCommandsToRun,
      upgradeCommandRunner.afterSyncMetadataUpgradeCommandsToRun,
      syncWorkspaceMetadataCommand.runOnWorkspace,
    ].forEach((fn) => expect(fn).toHaveBeenCalledTimes(1));

    // Verify order of execution
    const beforeSyncCall =
      beforeSyncMetadataUpgradeCommandsToRun.mock.invocationCallOrder[0];
    const afterSyncCall =
      afterSyncMetadataUpgradeCommandsToRun.mock.invocationCallOrder[0];
    const syncMetadataCall =
      syncWorkspaceMetadataCommand.runOnWorkspace.mock.invocationCallOrder[0];

    expect(beforeSyncCall).toBeLessThan(syncMetadataCall);
    expect(syncMetadataCall).toBeLessThan(afterSyncCall);
    expect(upgradeCommandRunner.migrationReport).toMatchInlineSnapshot();
  });

  it('should fail if workspace version is not the same as fromVersion', async () => {
    await buildModuleAndSetupSpies({
      numberOfWorkspace: 1,
      appVersion: '3.0.0',
      commandRunner: TestUpgradeCommandRunnerV2,
      workspaceOverride: {
        version: '0.1.0',
      },
    });

    const passedParams = [];
    const options = {};
    await upgradeCommandRunner.run(passedParams, options);
    expect(upgradeCommandRunner.migrationReport).toMatchInlineSnapshot(`
{
  "fail": [
    {
      "error": [Error: WORKSPACE_VERSION_MISSMATCH workspaceVersion=0.1.0 from=2.0.0 to=3.0.0],
      "workspaceId": "workspace_0",
    },
  ],
  "success": [],
}
`);
  });

  it('should fail if upgrade command version is invalid', async () => {
    await expect(
      buildModuleAndSetupSpies({
        commandRunner: InvalidVersionUpgradeCommandRunner,
      }),
    ).rejects.toThrowErrorMatchingInlineSnapshot(`"Invalid Version: invalid"`);
  });

  it('should fail if workspace version is not defined', async () => {
    await buildModuleAndSetupSpies({
      workspaceOverride: {
        version: null,
      },
    });

    const passedParams = [];
    const options = {};
    await upgradeCommandRunner.run(passedParams, options);
    expect(upgradeCommandRunner.migrationReport).toMatchInlineSnapshot();
  });

  it('should fail if app version is not defined', async () => {
    await buildModuleAndSetupSpies({
      appVersion: null,
    });

    const passedParams = [];
    const options = {};
    await expect(
      upgradeCommandRunner.run(passedParams, options),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Cannot run upgrade command when APP_VERSION is not defined "`,
    );
  });
});
