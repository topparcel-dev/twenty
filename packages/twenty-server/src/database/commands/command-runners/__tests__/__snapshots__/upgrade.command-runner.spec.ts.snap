// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpgradeCommandRunner Workspace upgrade should fail when APP_VERSION is not defined 1`] = `[Error: Cannot run upgrade command when APP_VERSION is not defined]`;

exports[`UpgradeCommandRunner Workspace upgrade should fail when workspace version is not defined 1`] = `[Error: WORKSPACE_VERSION_NOT_DEFINED to=2.0.0]`;

exports[`UpgradeCommandRunner Workspace upgrade should fail when workspace version is not equal to fromVersion 1`] = `[Error: WORKSPACE_VERSION_MISSMATCH Upgrade for workspace workspace_0 failed as its version is beneath fromWorkspaceVersion=2.0.0]`;

exports[`UpgradeCommandRunner should run upgrade command with failing and successful workspaces 1`] = `[Error: WORKSPACE_VERSION_MISSMATCH Upgrade for workspace outated_version_workspace failed as its version is beneath fromWorkspaceVersion=1.0.0]`;

exports[`UpgradeCommandRunner should run upgrade command with failing and successful workspaces 2`] = `[Error: Received invalid version: invalid 1.0.0]`;

exports[`UpgradeCommandRunner should run upgrade command with failing and successful workspaces 3`] = `[Error: WORKSPACE_VERSION_NOT_DEFINED to=2.0.0]`;
